### PR TITLE
Add basic Selenium tests for IQ# notebook

### PR DESCRIPTION
**Pull request by rmshaffer**
This PR adds basic Selenium-based browser integration tests to the IQ# CI pipeline. The tests validate that a Q# Jupyter Notebook can be opened in a browser, that a simple cell execution succeeds, and that the expected JavaScript modules are loaded.

### CODE REVIEW

**Comment by vxfield** [comment]
This is awesome, @rmshaffer!

I'm thinking what kind of unit tests we could create to test IQ# overall, but at the same time we have hundreds of cells ready to be tested in Katas and Samples repos that we could use as IQ# test cases.
Perhaps we could have a comprehensive test job that clones both Samples and Katas repos and run all of their cells?
What do you think?

**Comment by cgranade** [comment]
> Perhaps we could have a comprehensive test job that clones both Samples and Katas repos and run all of their cells?
> What do you think?

At the moment, we don't have any dependencies from the IQ# repo to the libraries, samples, or katas repos; that really helps prevent circular dependencies between repos. If we want to have that comprehensive test, I'd suggest isolating it in the e2e pipelines if possible.

**Comment by rmshaffer** [comment]
> Perhaps we could have a comprehensive test job that clones both Samples and Katas repos and run all of their cells?
What do you think?

I think it couldn't hurt to have something like this, @vxfield! In principle, the `jupyter nbconvert` tool already used by the Katas should give us pretty good confidence that all of the cells execute without throwing an error, but it doesn't tell us anything about whether the _output_ of the cells is correct. So I think that's where we can get particular value with Selenium - we could actually look at the output of cells that are encoded as HTML and/or rendered by client-side JS and verify that things look correct.

It may be that we get better return on investment by designing a few targeted cells to test - e.g., by exercising things like `DumpMachine()` that get special treatment in HTML. I'm planning to add some simple tests like this for the new visualization functionality that we are building in the `feature/visualization-improvements` branch.

Looking at the [build scripts in the QDK samples repo](https://github.com/microsoft/Quantum/tree/main/Build), it doesn't appear that there is currently any validation of notebooks in that repo at the moment. So maybe starting with a notebook there, e.g., the [Intro to IQ# notebook](https://github.com/microsoft/Quantum/blob/main/samples/getting-started/intro-to-iqsharp/Notebook.ipynb), might be interesting.

> If we want to have that comprehensive test, I'd suggest isolating it in the e2e pipelines if possible.

Agreed, @cgranade. At the very least, tests of any particular notebooks should probably live in the same repo as the notebooks themselves.

File: build/test-selenium.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+from notebook.notebookapp import list_running_servers
+from notebook.tests.selenium.utils import Notebook
+from selenium.common.exceptions import JavascriptException
+from selenium.webdriver import Firefox
+
+# workaround for missing os.uname attribute on some Windows systems
+if not hasattr(os, "uname"):
+    os.uname = lambda: ["Windows"]
+
+def setup_module():
```
Comment (anpaz) [inline build/test-selenium.py:14]: is this guaranteed to run in order?

**Review by anpaz — APPROVED** [review]

File: build/steps-selenium.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+steps:
+
+##
+# Pre-reqs
+##
+- pwsh: Write-Host "##vso[task.prependpath]$env:CONDA\Scripts"
+  displayName: "Add conda to PATH"
+
+- pwsh: conda create -n selenium-env -y python=3.7 pip setuptools pytest jupyter_client=6.1.7 notebook=6.1.1 selenium=3.141
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply build/steps-selenium.yml:14]: Not really, except that I need to install Python, Jupyter Notebook, Selenium, and Pytest, and the simplest way to do that seemed to be with a `conda create` command. I guess I could have also installed Python and used `pip install`... I'll see if that works.

File: build/ci.yml [hunk 3]
```diff
@@ -23,6 +23,13 @@ jobs:
     inputs:
       failOnAlert: true
 
+- job: "test_selenium"
```
[the remark this answers is not in the document]
Comment (rmshaffer) [reply build/ci.yml:26]: I agree. This should really only be for very basic browser/Jupyter integration testing, which is all internal to IQ#.

File: build/test-selenium.py (hunk 1 above)
Comment (rmshaffer) [reply build/test-selenium.py:14]: I think so. But I should just put this inside `setup_module()` anyway, it's cleaner that way.

**Review by anpaz — APPROVED** [review]

**Pull request merged**